File: A_PS/pythonPs/P_무인도 여행.py
from collections import deque
def solution(maps):
    r,c = len(maps), len(maps[0])
    result = []
    visited = [[False] * c for _ in range(r)]
    dx,dy = [-1, 1, 0, 0], [0, 0, -1, 1]
    
    def bfs(x,y):
        q = deque([(x,y)])
        visited[x][y] = True
        cnt = int(maps[x][y])
        while q:
            x,y = q.popleft()
            for i in range(4):
                nx,ny = x+dx[i], y+dy[i]
                if 0<=nx<r and 0<=ny<c and maps[nx][ny] != 'X' and not visited[nx][ny]:
                    visited[nx][ny] = True
                    cnt += int(maps[nx][ny])
                    q.append((nx,ny))
        return cnt
                    
    for i in range(r):
        for j in range(c):
            if maps[i][j] != 'X' and not visited[i][j]:
                result.append(bfs(i,j))
    
    if len(result) > 0:
        result.sort()
    else:
        result.append(-1)
    return result

solution(["XXX","XXX","XXX"])

File: A_PS/javaPs/P_무인도여행.java
package javaPs;

import java.util.*;

public class P_무인도여행 {
    static int[] dx = {-1, 1, 0, 0};
    static int[] dy = {0, 0, -1, 1};
    static boolean[][] visited;
    public static void main(String[] args) {
        solution(new String[]{"X591X","X1X5X","X231X", "1XXX1"});
    }


    public static List<Integer> solution(String[] maps) {
        int[] answer = {};
        List<Integer> list = new ArrayList<>();

        visited = new boolean[maps.length][maps[0].length()];
        String[][] graph = new String[maps.length][maps[0].length()];


        for(int i=0; i<maps.length; i++) {
            graph[i] =  maps[i].split("");
        }

        System.out.println(Arrays.deepToString(graph));

        for(int i=0; i<maps.length; i++) {
            for(int j=0; j<maps[0].length(); j++) {
                if(!graph[i][j].equals("X") && !visited[i][j]) {
                    list.add(bfs(graph, i, j));
                }
            }
        }

        if(list.isEmpty()) list.add(-1);
        Collections.sort(list);



        return list;
    }

    public static int bfs(String[][] graph,int x,int y) {
        Queue<int[]> q = new LinkedList<>();
        int cnt=0;
        q.offer(new int[]{x,y});
        visited[x][y] = true;

        while (!q.isEmpty()) {
            int[] polled = q.poll();
            x = polled[0];
            y = polled[1];
            cnt += Integer.parseInt(graph[x][y]);

            for(int i=0; i<4; i++) {
                int nx = x + dx[i];
                int ny = y + dy[i];

                if (nx<0 || ny<0 || nx >= graph.length || ny >= graph[0].length) continue;
                if (!visited[nx][ny] && !graph[nx][ny].equals("X")) {
                    visited[nx][ny] = true;
                    q.offer(new int[]{nx, ny});
                }
            }
        }
        return cnt;
    }
}
